pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -1200,4 +1196,111 @@ public String setNamespaceBundleAffinity(String bundle, String broker) {
+            NamespaceBundle bundle = pulsar.getNamespaceService().getNamespaceBundleFactory()
+                    .getBundle(namespaceName, bundleRange);
+            namespaceLocalBundleData.computeIfAbsent(namespaceName, (key) -> new HashSet<>()).add(bundle);
+        }
+
+        Set<NamespaceBundle> realClusterBundles = new HashSet<>();
+
+        // check each namespace if bundle is valid in the cluster.
+        for (Map.Entry<String, Set<NamespaceBundle>> namespaceEntry : namespaceLocalBundleData.entrySet()) {
+            if (!isLeader()) {
+                return;
+            }
+
+            // check if namespace exist.
+            NamespaceName namespace = NamespaceName.get(namespaceEntry.getKey());
+            Set<NamespaceBundle> loadManagerKnownBundles = namespaceEntry.getValue();
+
+            Optional<Policies> nsPolicies;
+
+            try {
+                nsPolicies = pulsar.getPulsarResources()
+                        .getNamespaceResources().getPolicies(namespace);
+            } catch (MetadataStoreException e) {
+                log.error("error when get policies for namespace {}", namespace, e);
+                continue;
+            }
+
+            if (nsPolicies.isEmpty()) {
+                loadManagerKnownBundles.forEach((bundle) -> {
+                    String bundleString = bundle.toString();
+                    loadManagerBundleData.remove(bundleString);
+                    if (isLeader()) {
+                        log.info("namespace [{}] is not exist in cluster, "
+                                + "remove load-balance bundle-data {} from metadata store.", namespace, bundleString);
+                        deleteBundleDataFromMetadataStore(bundleString);
+                    }
+                });
+
+                continue;
+            }
+
+            // namespace is being deleted, do nothing because delete namespace will handle the bundle-data deletion.
+            if (nsPolicies.get().deleted) {
+                if (log.isDebugEnabled()) {
+                    log.debug("ignore namespace {} because the namespace is mark deleted", namespace);
+                }
+                continue;
+            }
+
+            NamespaceBundles namespaceBundles =
+                    pulsar.getNamespaceService()
+                            .getNamespaceBundleFactory()
+                            .getBundles(namespace);
+
+            if (log.isDebugEnabled()) {
+                log.debug("bundles in namespace {} is {}", namespace, namespaceBundles.getBundles());
+            }
+
+            if (namespaceBundles == null) {
+                continue;
+            }
+
+            realClusterBundles.clear();
+            realClusterBundles.addAll(namespaceBundles.getBundles());
+
+            Sets.SetView<NamespaceBundle> alreadyNotExistBundles =
+                    Sets.difference(loadManagerKnownBundles, realClusterBundles);
+
+            alreadyNotExistBundles.forEach((bundle) -> {
```

Comment:
Important to clarify:
I think this will only remove parent bundles, but we already remove the parent bundle right after the split like the one below.
Also,  this will not remove bundles that become inactive(no traffic), which is different from the existing behavior.

```
                    this.pulsar.getNamespaceService().getNamespaceBundleFactory()
                            .invalidateBundleCache(NamespaceName.get(namespaceName));
                    deleteBundleDataFromMetadataStore(bundleName);
```

Reply:
Yes, there are two way to trigger the bundle split. 
1. The LoadManager leader to trigger bundle split task.
2. The admin api, which is execute by the namespace bundle owner. the bundle owner is not deleteBundleDataFrom metadataStore.

the case two is covered here to cleanup.

Reply:
the bundle-data is exist all the time when the bundle is valid in the cluster, this is the behavior as always.
more background is in issue https://github.com/apache/pulsar/issues/13973 which try to remove the split parent bundle.

Reply:
I still would like to clarify this.

> Also, this will not remove bundles that become inactive(no traffic), which is different from the existing behavior.

Reply:
emmmm, sorry i did not get your point. In my view for current code or fix before https://github.com/apache/pulsar/issues/13973 there is no logic to delete no traffic bundles. If we need this feature, maybe in another patch to add it. By the way, what kind of bundle do we think there is no traffic ？and no we need to delete no traffic bundles？

Reply:
To clarify, https://github.com/apache/pulsar/pull/13974 commit introduced this logic, "remove inactive(no-traffic in `statsMap`) bundles" by `bundleData.remove(bundle); and deleteBundleDataFromMetadataStore(bundle);` I would like to confirm if we are fine with removing this logic from the backward-compatibility point of view.

> By the way, what kind of bundle do we think there is no traffic ？and do we need to delete no traffic bundles？

For example, when all topics in bundle x suddenly receive no traffic(all pub/sub-sessions are closed), the logic removes that bundle from the in-memory cache and metadata store.

Reply:
@heesung-sn hi, I create a standalone instance and enable ModularLoadMangerImpl the result show even all sessions are closed the current master won't clear the bundle-data. the result can be see at https://gist.github.com/lifepuzzlefun/d1181ed0eb97c82229298ce69b31daea

first create a test topic (non-partitioned) and start perf producer. we can see from the bundle stats there are incoming traffic. after close all the client to this topic, no traffic on this bundle 0x40000000_0x80000000. and after about 2 hours later. the bundle data is still here. and also bundle 0xc0000000_0xffffffff is always no traffic the loadmanager is not delete them

Reply:
i my opinion. the **active** in https://github.com/apache/pulsar/pull/13974 only means 

> clean inactive bundle from bundleData in loadData and bundlesCache after the bundle be split or unload

not no traffic. if we have logic to clear up no traffic bundle-data  the code has logic to verify if the bunde has traffic.

or if the logic is just rely on the load report to think if the bundle-data is not reported. the bundle is no traffic at all. there should be logic to generate bundleMetric to check if the bundle has no logic the load-report should not contain this bundle. but both logic did not occur on the master.

And the if we think the current logic as a clean up for the no-traffic bundle-data. I just have to say this is a bug, not a feature. because in my production environment has 10000 bundles with traffic, only less than 100 bundle-data on the metadata-store all the time, because of the clean up logic.

Reply:
I guess inactive bundles are cleaned only after unload(or split), when the unload or split forcefully cleans the bundle in the statsMap and other cache/stores.

To confirm, can you try to unload the bundle after stopping the traffic in your test and see if they are indeed cleaned only after unload?

If the above is true, when the next lookup happened on that unloaded bundle, it should've created the bundle back in the bundle cache and metadata store. I dont understand why you didt see this in your environment. I wonder if we need to dig more why the bundles didn't get restored in the cache when they were reactivated.